src/ra_server.erl

```diff
@@ -2803,6 +2850,62 @@ meta_name(#cfg{system_config = #{names := #{log_meta := Name}}}) ->
     Name;
 meta_name(#{names := #{log_meta := Name}}) ->
     Name.
+
+%%% ====================
+%%% Voter status helpers
+%%% ====================
+
+-spec new_nonvoter(ra_server_state()) -> ra_voter_status().
+
+new_nonvoter(#{commit_index := Target} = _State) ->
+    {no, #{target => Target}}.
+
+-spec maybe_promote_voter(ra_server_id(), ra_server_state(), effects()) -> effects().
+
+maybe_promote_voter(PeerID, #{cluster := Cluster} = _State, Effects) ->
+    % Unknown peer handled in the caller.
+    #{PeerID := #{match_index := MI, voter := OldStatus}} = Cluster,
+    case update_voter_status(OldStatus, MI) of
+        OldStatus ->
+            Effects;
+        yes ->
+            [{next_event,
+                {command, {'$ra_join',
+                    #{ts => os:system_time(millisecond)},
+                    #{node => PeerID, voter => yes},
+                noreply}}} |
+             Effects]
+    end.
+
+update_voter_status({no, #{target := Target}}, MI)
+  when MI >= Target ->
+    yes;
+update_voter_status(Permanent, _) ->
+    Permanent.
+
+-spec voter_status(ra_server_id(), ra_cluster()) -> ra_voter_status().
+
+voter_status(PeerId, Cluster) ->
+    case maps:get(PeerId, Cluster, undefined) of
+        undefined ->
+            {no, undefined};
+        Peer ->
+            maps:get(voter, Peer, yes)
+    end.
+
+-spec required_quorum(ra_cluster()) -> pos_integer().
+
+required_quorum(Cluster) ->
+    Voters = count_voters(Cluster),
+    trunc(Voters / 2) + 1.
+
+count_voters(Cluster) ->
+    maps:fold(
+      fun (_, #{voter := {no, _}}, Count) -> Count;
+          (_, _, Count)                   -> Count + 1
```

Comment:
Use the same formatting in the rest of the code base (i.e. remove redunantant indentation before `->` and favour putting the expression that follows on a new line. :)